hazelcast-jet-core/src/main/java/com/hazelcast/jet/aggregate/AggregateOperations.java

```diff
@@ -233,54 +239,159 @@ private AggregateOperations() {
+            @Nonnull DistributedTriFunction<? super R1, ? super R2, ? super R3, R> finishFn
+    ) {
+        return AggregateOperation
+                .withCreate(() -> tuple3(op1.createFn().get(), op2.createFn().get(), op3.createFn().get()))
+                .<T>andAccumulate((acc, item) -> {
+                    op1.accumulateFn().accept(acc.f0(), item);
+                    op2.accumulateFn().accept(acc.f1(), item);
+                    op3.accumulateFn().accept(acc.f2(), item);
                 })
-                .andCombine(
-                        // we can support combine only if all operations do
-                        Stream.of(untypedOps).allMatch(o -> o.combineFn() != null)
-                                ? (accs1, accs2) -> {
-                                    for (int i = 0; i < untypedOps.length; i++) {
-                                        untypedOps[i].combineFn().accept(accs1.get(i), accs2.get(i));
-                            }
-                        } : null)
-                .andDeduct(
-                        // we can support deduct only if all operations do
-                        Stream.of(untypedOps).allMatch(o -> o.deductFn() != null)
-                                ? (accs1, accs2) -> {
-                                    for (int i = 0; i < untypedOps.length; i++) {
-                                        untypedOps[i].deductFn().accept(accs1.get(i), accs2.get(i));
-                                    }
-                                }
-                                : null)
-                .andFinish(accs -> {
-                    List<Object> res = new ArrayList<>(untypedOps.length);
-                    for (int i = 0; i < untypedOps.length; i++) {
-                        res.add(untypedOps[i].finishFn().apply(accs.get(i)));
-                    }
-                    return res;
-                });
+                .andCombine(op1.combineFn() == null || op2.combineFn() == null ? null :
+                        (acc1, acc2) -> {
+                            op1.combineFn().accept(acc1.f0(), acc2.f0());
+                            op2.combineFn().accept(acc1.f1(), acc2.f1());
+                            op3.combineFn().accept(acc1.f2(), acc2.f2());
+                        })
+                .andDeduct(op1.deductFn() == null || op2.deductFn() == null ? null :
+                        (acc1, acc2) -> {
+                            op1.deductFn().accept(acc1.f0(), acc2.f0());
+                            op2.deductFn().accept(acc1.f1(), acc2.f1());
+                            op3.deductFn().accept(acc1.f2(), acc2.f2());
+                        })
+                .andFinish(acc -> finishFn.apply(
+                        op1.finishFn().apply(acc.f0()),
+                        op2.finishFn().apply(acc.f1()),
+                        op3.finishFn().apply(acc.f2())));
+    }
+
+    /**
+     * Returns a builder to create a composite aggregate operation, if the
```

Comment:
it might be better to just say this can be used with an arbitrary number of operations and then add a note to refer to one of the two or three-arity ones. If you have less than 4, you can still use this option (for whatever reason).

Reply:
fixed